docs/src/docs/asciidoc/end-to-end-gradle-guide.adoc

```diff
@@ -0,0 +1,871 @@
+
+Here is an example of a valid `metadataCopy` usage:
+
+[source,bash,subs="verbatim,attributes", role="multi-language-sample"]
+----
+./gradlew metadataCopy --task test --dir resources/META-INF/native-image/org.example
+----
+
+[WARNING]
+====
+Note that **if you store generated metadata on location other than the default one**, you will need to pass that location as a Native Image `buildArg` with `-H:ConfigurationFileDirectories` option.
+For example, you can pass that argument inside the `binaries.all` (or `binaries.test` or `binaries.main` depending on your use-case) block like this:
+
+[source,groovy,subs="verbatim,attributes", role="multi-language-sample"]
+----
+binaries.all {
+    buildArgs.add("-H:ConfigurationFileDirectories=path/to/metadata")
+}
+----
+
+[source,kotlin,subs="verbatim,attributes", role="multi-language-sample"]
+----
+binaries.all {
+    buildArgs.add("-H:ConfigurationFileDirectories=path/to/metadata")
+}
+----
+====
+
+[[additional-agent-options]]
+=== Additional Native Image Agent options
+
+As your project grows, you should consider configuring the agent to gain more control over the generated metadata.
+
+First thing that you can configure is the agent mode.
+There are three possible agent modes:
+
+* `standard` - only generates metadata without any special processing (this is the default mode). No additional options available.
+* `conditional` - entries of the generated metadata will be included in the Native Image only if the condition in the entry is satisfied. Consumes following options:
+** `userCodeFilterPath` - specifies a filter file used to classify classes as user application classes. Generated conditions will only reference these classes See <<agent-filter-file, the following section>>
+** `extraFilterPath` - (optional) extra filter used to further filter the collected metadata. See <<agent-filter-file, the following section>>
+* `direct` - in this mode user configures the agent completely manually by adding all options with:
+** `options.add("<option>")`
+
+Each of the described modes has its own benefits.
+For example:
+
+- `standard` mode is a **great starting point** in your project development
+- `conditional` mode is mainly aimed towards **library maintainers** with the goal of reducing overall footprint
+- `direct` mode is for **experienced users** that knows how to configure the agent manually
+
+You can configure each mode (and declare the one that will be used for generating metadata) inside the `agent` block in `build.gradle` file.
+Here is an example of the `agent` block with configured conditional and direct modes, where the conditional mode is set as default and will be used to generate the metadata:
+
+[source,groovy,subs="verbatim,attributes", role="multi-language-sample"]
+----
+agent {
+    enabled = true
+    defaultMode = "conditional"
+    modes {
+        conditional {
+            userCodeFilterPath = "src/test/resources/metadata.with.starting.user-code-filter/user-code-filter.json"
```

Comment:
:nit: Technically speaking, files found in `src/test/resources` are found on classpath at runtime (here, for tests). This doesn't quite make sense to have the user code filter at runtime, so I would personally use a different directory, for example `src/test/native-image/...`